In [ ]:
from google.colab import files


uploaded = files.upload()

In [2]:
import pandas as pd  # Data manipulation and analysis
import numpy as np  # Numerical operations
import matplotlib.pyplot as plt  # Data visualization
import os
import io

cab_data = pd.read_csv(io.BytesIO(uploaded['Cab_Data.csv']))
city_data = pd.read_csv(io.BytesIO(uploaded['City.csv']))
tran_data = pd.read_csv(io.BytesIO(uploaded['Transaction_ID.csv']))
cust_data = pd.read_csv(io.BytesIO(uploaded['Customer_ID.csv']))

In [ ]:
data1 = pd.merge(tran_data, cust_data, on="Customer ID")
print(data1)
data2 = pd.merge(data1, cab_data, on="Transaction ID")
print(data2)
comb_data = pd.merge(data2, city_data, on="City")
print(comb_data)

In [ ]:
comb_data.info()

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

# Select numerical variables
numerical_vars = comb_data.select_dtypes(include='number')

# Create individual boxplots for each numerical variable
for column in numerical_vars.columns:
    plt.figure(figsize=(8, 6))
    sns.boxplot(data=comb_data[column])
    plt.title(f'Box Plot of {column}')
    plt.show()

In [ ]:
descriptive_stats = numerical_vars.describe()
print(descriptive_stats)

In [ ]:
for column in comb_data.columns:
  if comb_data[column].dtype == 'object':
    categorical_freq = comb_data[column].value_counts()
    print(categorical_freq)

In [ ]:
#univariate visualizations for numerical variables
for column in comb_data.columns:
  if comb_data[column].dtype == 'float64' or comb_data[column].dtype == 'int64':
    plt.hist(comb_data[column], bins = 10)
    plt.xlabel(column)
    plt.ylabel('frequency')
    plt.title(f'Histograms of {column}')
    plt.show()

In [ ]:
#univariate visualizations for categorical variables
for column in comb_data.columns:
    # Check if the column has a categorical data type
    if comb_data[column].dtype == 'object':
        # Generate a count plot (bar plot) for the categorical variable
        plt.figure(figsize=(10, 6))
        sns.countplot(data=comb_data, x=column)
        plt.xlabel(column)
        plt.ylabel('Count')
        plt.title(f'Count Plot of {column}')
        plt.xticks(rotation=45)
        plt.show()

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

plt.figure(figsize=(8, 6))
sns.countplot(data=comb_data, x='Gender', hue='Company')
plt.xlabel('Gender')
plt.ylabel('Count')
plt.title('Stacked Bar Plot of Gender and Company')
plt.show()

In [ ]:
plt.figure(figsize = (6,5))
sns.boxplot(data=comb_data, x='Company', y='Age')
plt.xlabel('Company')
plt.ylabel('Age')
plt.title('Age stats for companies')
plt.show()

In [ ]:
plt.figure(figsize=(5,4))
sns.boxplot(data=comb_data, x='Company', y='Income (USD/Month)')
plt.xlabel('Company')
plt.ylabel('Income (USD/Month)')
plt.title('Customer income stats for companies')
plt.show()

In [ ]:
plt.figure(figsize=(25, 6))
sns.countplot(data=comb_data, x='City', hue='Company')
plt.xlabel('City')
plt.ylabel('Count')
plt.title('City vs Company')
plt.show()

In [ ]:
plt.figure(figsize=(8,8))
sns.scatterplot(data=comb_data, x='Price Charged', y='KM Travelled')
plt.xlabel('Price Charged')
plt.ylabel('KM Travelled')
plt.title('price vs km')
plt.show()

In [ ]:
plt.figure(figsize=(8,8))
sns.boxplot(data=comb_data, x='Company', y='Cost of Trip')
plt.xlabel('Compnay')
plt.ylabel('Cost of Trip')
plt.title('company vs cost')
plt.show()

In [20]:
# Remove commas from the Users column
company1_data = comb_data[comb_data['Company'] == 'Pink Cab']
company2_data = comb_data[comb_data['Company'] == 'Yellow cab']

from google.colab import files
company1_data.to_csv('company1_data.csv')
files.download('company1_data.csv')
company2_data.to_csv('company2_data.csv')
files.download('company2_data.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
from google.colab import files


uploaded = files.upload()

In [ ]:
#trends of users over time

company1_data = pd.read_csv('company1_data (1).csv')
company2_data = pd.read_csv('company2_data (1).csv')

company1_data['Users'] = company1_data['Users'].str.replace(',', '')
company2_data['Users'] = company2_data['Users'].str.replace(',', '')

# Convert the Users column to numeric type
company1_data['Users'] = pd.to_numeric(company1_data['Users'])
company2_data['Users'] = pd.to_numeric(company2_data['Users'])

# Convert date column to datetime
company1_data['Date of Travel'] = pd.to_datetime(company1_data['Date of Travel'])
company2_data['Date of Travel'] = pd.to_datetime(company2_data['Date of Travel'])

# Set date column as index
company1_data.set_index('Date of Travel', inplace=True)
company2_data.set_index('Date of Travel', inplace=True)

# Resample the data by month and calculate the average number of customers
company1_resampled = company1_data['Users'].resample('M').mean()
company2_resampled = company2_data['Users'].resample('M').mean()

# Plot the seasonality and trends
plt.figure(figsize= (15,6))
plt.plot(company1_resampled.index, company1_resampled, label='Pink Cab')
plt.plot(company2_resampled.index, company2_resampled, label='Yellow Cab')
plt.xlabel('Date')
plt.ylabel('Average Number of Customers')
plt.title('Seasonality of Cab Customers')
plt.legend()
plt.show()

In [ ]:
plt.figure(figsize=(25, 6))
sns.countplot(data=comb_data, x='Company', hue='Users')
plt.xlabel('City')
plt.ylabel('Count')
plt.title('City vs Company')
plt.show()

In [ ]:
comb_data['profit'] = comb_data['Price Charged'] - comb_data['Cost of Trip']
print(comb_data)

In [ ]:
plt.figure(figsize=(15, 6))
plt.scatter(data = comb_data, x='Users', y='profit')
plt.xlabel('Number of Customers')
plt.ylabel('Profit')
plt.title('Relationship between Number of Customers and Profit')
plt.show()

In [34]:
from google.colab import files
comb_data.to_csv('comb_data.csv')
files.download('comb_data.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [35]:
from google.colab import files


uploaded = files.upload()

Saving comb_data (2).csv to comb_data (2).csv


In [38]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

# Load the data
data = pd.read_csv(io.BytesIO(uploaded['comb_data (2).csv']))

# Select the relevant features and target variable
categorical_features = ['Payment_Mode', 'Gender', 'Company', 'City', 'Date of Travel']  # Replace with actual categorical feature names
numerical_features = ['Age', 'Income (USD/Month)', 'KM Travelled', 'Cost of Trip']  # Replace with actual numerical feature names
target = 'Price Charged'

# Convert categorical features to numerical using label encoding
label_encoder = LabelEncoder()
for feature in categorical_features:
    data[feature] = label_encoder.fit_transform(data[feature])

# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(data[numerical_features + categorical_features], data[target], test_size=0.2, random_state=42)

# Add a constant term to the features
X_train = sm.add_constant(X_train)

# Create and fit the linear regression model
model = sm.OLS(y_train, X_train)
results = model.fit()

# Print the regression summary
print(results.summary())

# Predict the target variable for the test set
X_test = sm.add_constant(X_test)
y_pred = results.predict(X_test)

# Evaluate the model performance
mse = mean_squared_error(y_test, y_pred)
print('Mean Squared Error:', mse)

                            OLS Regression Results                            
Dep. Variable:          Price Charged   R-squared:                       0.757
Model:                            OLS   Adj. R-squared:                  0.757
Method:                 Least Squares   F-statistic:                 9.936e+04
Date:                Wed, 21 Jun 2023   Prob (F-statistic):               0.00
Time:                        08:23:19   Log-Likelihood:            -1.8194e+06
No. Observations:              287513   AIC:                         3.639e+06
Df Residuals:                  287503   BIC:                         3.639e+06
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                -92.0016      1

In [39]:
#model without including 'Payment_Mode'
# Select the relevant features and target variable
categorical_features = ['Gender', 'Company', 'City', 'Date of Travel']  # Replace with actual categorical feature names
numerical_features = ['Age', 'Income (USD/Month)', 'KM Travelled', 'Cost of Trip']  # Replace with actual numerical feature names
target = 'Price Charged'

# Convert categorical features to numerical using label encoding
label_encoder = LabelEncoder()
for feature in categorical_features:
    data[feature] = label_encoder.fit_transform(data[feature])

# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(data[numerical_features + categorical_features], data[target], test_size=0.2, random_state=42)

# Add a constant term to the features
X_train = sm.add_constant(X_train)

# Create and fit the linear regression model
model = sm.OLS(y_train, X_train)
results = model.fit()

# Print the regression summary
print(results.summary())

# Predict the target variable for the test set
X_test = sm.add_constant(X_test)
y_pred = results.predict(X_test)

# Evaluate the model performance
mse = mean_squared_error(y_test, y_pred)
print('Mean Squared Error:', mse)

                            OLS Regression Results                            
Dep. Variable:          Price Charged   R-squared:                       0.757
Model:                            OLS   Adj. R-squared:                  0.757
Method:                 Least Squares   F-statistic:                 1.118e+05
Date:                Wed, 21 Jun 2023   Prob (F-statistic):               0.00
Time:                        08:24:12   Log-Likelihood:            -1.8194e+06
No. Observations:              287513   AIC:                         3.639e+06
Df Residuals:                  287504   BIC:                         3.639e+06
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                -91.8401      1

**Hypotheses analysed:**

1.   Which gender prefers which cab company most
2.   Average age of customers - company wise
3.   Income stats of the customers - company wise
4.   which cab company is predominant in which city
5.   the price charged vs km travelled
6.   the cost of trip stats - company wise

**Observations:**

*   In both males and females, yellow cab company is predominant, from barplot
*   The mean age of the customers for both the cab compnaies is equal, from the boxplot
*   The average customer income for both pink and yellow cab companies is equal, from the boxplot (~15000 USD)
*   The Yellow cab company is predominant among majprity of the cities, except a very few. From the countplot.
*   The sctter plot of price charged vs KM travelled has a linear trend (positive), which shows for increase in either price charged or km travelled, there is a similar effect on the other variable.
*   the mean of cost of trip is higher for yellow cab company when compared to the pink cab company. This provides strong evidence for XYZ to invest in yellow cab company.
*   From the analysis of seasonality in cab customers over date of travel, the graph shows a similar trnd for both companies over time

**Analysis:**

Conducted a linear regression for predicting the price charged for each trip for both the pink and yellow cab companies. the regression model had R^2 value of 0.757, which shows the generated model is fit for prediction. The first model (which includes Payment_mode in the analysis) has a better MSE value of 18262.3514

**Conclusion:**

My recommendation for XYZ would be to invest in the yellow cab company from the abive analyses and observations.


